pkg/controller/master/node/node_volume_detach_controller.go

```diff
@@ -0,0 +1,190 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	ctlcorev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/harvester/harvester/pkg/config"
+	ctllonghornv1 "github.com/harvester/harvester/pkg/generated/controllers/longhorn.io/v1beta1"
+	ctlsnapshotv1 "github.com/harvester/harvester/pkg/generated/controllers/snapshot.storage.k8s.io/v1beta1"
+	"github.com/harvester/harvester/pkg/indexeres"
+	"github.com/harvester/harvester/pkg/util"
+)
+
+const (
+	nodeVolumeDetachControllerName  = "node-volume-detach-controller"
+	nodeVolumeDetachEnqueueInterval = 5 * time.Second
+)
+
+// nodeVolumeDetachController is responsible for detaching volumes from a node when the node is in draining
+// fix https://github.com/harvester/harvester/issues/3681
+type nodeVolumeDetachController struct {
+	nodeController ctlcorev1.NodeController
+	podCache       ctlcorev1.PodCache
+	pvcCache       ctlcorev1.PersistentVolumeClaimCache
+	volumes        ctllonghornv1.VolumeClient
+	volumeCache    ctllonghornv1.VolumeCache
+	snapshotCache  ctlsnapshotv1.VolumeSnapshotCache
+}
+
+// VolumeDetachRegister registers the node volume detach controller
+func VolumeDetachRegister(ctx context.Context, management *config.Management, options config.Options) error {
+	nodes := management.CoreFactory.Core().V1().Node()
+	pods := management.CoreFactory.Core().V1().Pod()
+	pvcs := management.CoreFactory.Core().V1().PersistentVolumeClaim()
+	volumes := management.LonghornFactory.Longhorn().V1beta1().Volume()
+	snapshots := management.SnapshotFactory.Snapshot().V1beta1().VolumeSnapshot()
+	nodeVolumeDetachController := &nodeVolumeDetachController{
+		nodeController: nodes,
+		podCache:       pods.Cache(),
+		pvcCache:       pvcs.Cache(),
+		volumes:        volumes,
+		volumeCache:    volumes.Cache(),
+		snapshotCache:  snapshots.Cache(),
+	}
+
+	nodes.OnChange(ctx, nodeVolumeDetachControllerName, nodeVolumeDetachController.OnNodeChanged)
+
+	return nil
+}
+
+// OnNodeChanged is called when a node is changed
+func (c *nodeVolumeDetachController) OnNodeChanged(key string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.DeletionTimestamp != nil {
+		return node, nil
+	}
+
+	// skip if node is not in draining
+	if !isNodeInDraining(node) {
+		return node, nil
+	}
+	logrus.Debugf("node %s is in draining, try to detach unused volumes", node.Name)
+
+	// get all longhorn volume attach to this node
+	volumes, err := c.volumeCache.GetByIndex(indexeres.VolumeByNodeIndex, node.Name)
+	if err != nil {
+		return node, err
+	}
+	waitPVCs := make([]*corev1.PersistentVolumeClaim, 0, len(volumes))
+	for _, volume := range volumes {
```

Comment:
Need to exclude `detached` / filter `attached` volumes  at first, no need to check all volumes in following steps.

```
harv2:~ # kk get volume -A
NAMESPACE         NAME                                       STATE      ROBUSTNESS   SCHEDULED   SIZE          NODE    AGE
longhorn-system   pvc-6be2504c-dc80-4824-a8ec-8d19ec3b2e0b   attached   degraded                 5368709120    harv2   12d
longhorn-system   pvc-6c6a5088-38c1-41f0-b3f9-7b7e8668fb59   attached   degraded                 53687091200   harv2   12d
longhorn-system   pvc-c5083f9c-cb04-4a5c-8869-8a2d3bb23d86   attached   degraded                 2147483648    harv2   12d


longhorn-system   pvc-f160a980-0d0e-4a48-bfed-7099e46e3b3e   detached   unknown                  1073741824            62m
```